From: Thomas Renninger <trenn@suse.de>
Subject: intel_idle: Add a boot param to keep flushing TLBs
Patch-Mainline: not yet
References: bnc#675161

This is a test patch. It may show up mainline if it helps, otherwise
it can get reverted again soon.

---
 drivers/idle/intel_idle.c |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

Index: linux-2.6.38-master/drivers/idle/intel_idle.c
===================================================================
--- linux-2.6.38-master.orig/drivers/idle/intel_idle.c
+++ linux-2.6.38-master/drivers/idle/intel_idle.c
@@ -91,6 +91,7 @@ static struct cpuidle_state *cpuidle_sta
  */
 static unsigned long long auto_demotion_disable_flags;
 
+static int tlb_flushing;
 /*
  * Set this flag for states where the HW flushes the TLB for us
  * and so we don't need cross-calls to keep it consistent.
@@ -229,7 +230,7 @@ static int intel_idle(struct cpuidle_dev
 	 * leave_mm() to avoid costly and often unnecessary wakeups
 	 * for flushing the user TLB's associated with the active mm.
 	 */
-	if (state->flags & CPUIDLE_FLAG_TLB_FLUSHED)
+	if (!tlb_flushing && state->flags & CPUIDLE_FLAG_TLB_FLUSHED)
 		leave_mm(cpu);
 
 	if (!(lapic_timer_reliable_states & (1 << (cstate))))
@@ -510,7 +511,8 @@ module_init(intel_idle_init);
 module_exit(intel_idle_exit);
 
 module_param(max_cstate, int, 0444);
-module_param(lapic_timer_reliable_states, int, 0444);
+module_param(lapic_timer_reliable_states, uint, 0444);
+module_param(tlb_flushing, int, 0644);
 
 MODULE_AUTHOR("Len Brown <len.brown@intel.com>");
 MODULE_DESCRIPTION("Cpuidle driver for Intel Hardware v" INTEL_IDLE_VERSION);
